src/psyclone/nemo.py

```diff
@@ -0,0 +1,822 @@
+        prog_unit = self.root.invoke._ast
+        spec = walk_ast(prog_unit.content, [Fortran2003.Specification_Part])
+        if not spec:
+            names = walk_ast(prog_unit.content, [Fortran2003.Name])
+            # TODO create a Specification_Part
+            raise InternalError("No specification part found for routine {0}!".
+                                format(names[0]))
+        # TODO check that this variable has not already been declared
+        # Requires that we capture all variable declarations in the routine
+        # in some sort of management class.
+        decln = Fortran2003.Type_Declaration_Stmt(
+            FortranStringReader("integer :: {0}".format(self._variable_name)))
+        spec[0].content.append(decln)
+
+        for subsec in subsections:
+            # A tuple is immutable so work with a list
+            indices = list(subsec.items)
+            # Replace the colon with our new variable name
+            indices[outermost_dim] = name
+            # Replace the original tuple with a new one
+            subsec.items = tuple(indices)
+
+        # Create an explicit loop
+        text = '''\
+do {0}=1,{1},{2}
+  replace = me
+end do
+'''.format(self._variable_name, self._stop, self._step)
+        loop = Fortran2003.Block_Nonlabel_Do_Construct(
+            FortranStringReader(text))
+        parent_index = ast._parent.content.index(ast)
+        # Insert it in the fparser2 AST at the location of the implicit
+        # loop
+        ast._parent.content.insert(parent_index, loop)
+        # Replace the content of the loop with the (modified) implicit
+        # loop
+        loop.content[1] = ast
+        # Remove the implicit loop from its original parent in the AST
+        ast._parent.content.remove(ast)
+        # Update the parent of the AST
+        ast._parent = loop
+        # Update our own pointer into the AST to now point to the explicit
+        # loop we've just created
+        self._ast = loop
+
+        if NemoImplicitLoop.match(ast):
+            # We still have an implicit loop so recurse
+            self.addchild(NemoImplicitLoop(ast, parent=self))
+        else:
+            # We must be left with a kernel
+            self.addchild(NemoKern(ast, parent=self))
+
+    @staticmethod
+    def match(node):
+        '''
+        Checks whether the supplied node in the fparser2 AST represents
+        an implicit loop (using Fortran array syntax).
+        :param node: node in the fparser2 AST to check
+        :type node: :py:class:`fparser.two.Fortran2003.Assignment_Stmt`
+        :returns: True if the node does represet an implicit loop.
+        :rtype: bool
+        '''
+        if not isinstance(node, Fortran2003.Assignment_Stmt):
+            return False
+        # We are expecting something like:
+        #    array(:,:,jk) = some_expression
+        # so we check the left-hand side...
+        lhs = node.items[0]
+        if isinstance(lhs, Fortran2003.Part_Ref):
```

Comment:
Again I think that whether a statement uses array syntax or not should be provided by a library.